### PR TITLE
in stdin: add buffer_size parameter

### PR DESCRIPTION
**Pull request by dojci**
Add ability to configure buffer size for stdin input plugin with respect of default value (2 * 8192)

Addresses #2199 #89

----
**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [ ] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by dojci** [comment]
Example configuration file:
```
[INPUT]
    Name           stdin
    Buffer_size    6MB
[Output]
    Name           stdout
```

**Comment by dojci** [comment]
Default buffer size:
```
[SERVICE]
    Log_Level       debug
    Daemon          Off
[INPUT]
    Name           stdin
[Output]
    Name           stdout

```
```
Fluent Bit v1.6.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/07/17 00:22:37] [ info] Configuration:
[2020/07/17 00:22:37] [ info]  flush time     | 5.000000 seconds
[2020/07/17 00:22:37] [ info]  grace          | 5 seconds
[2020/07/17 00:22:37] [ info]  daemon         | 0
[2020/07/17 00:22:37] [ info] ___________
[2020/07/17 00:22:37] [ info]  inputs:
[2020/07/17 00:22:37] [ info]      stdin
[2020/07/17 00:22:37] [ info] ___________
[2020/07/17 00:22:37] [ info]  filters:
[2020/07/17 00:22:37] [ info] ___________
[2020/07/17 00:22:37] [ info]  outputs:
[2020/07/17 00:22:37] [ info]      stdout.0
[2020/07/17 00:22:37] [ info] ___________
[2020/07/17 00:22:37] [ info]  collectors:
[2020/07/17 00:22:37] [ info] [engine] started (pid=2472338)
[2020/07/17 00:22:37] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/07/17 00:22:37] [debug] [storage] [cio stream] new stream registered: stdin.0
[2020/07/17 00:22:37] [ info] [storage] version=1.0.4, initializing...
[2020/07/17 00:22:37] [ info] [storage] in-memory
[2020/07/17 00:22:37] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/07/17 00:22:37] [debug] [input:stdin:stdin.0] buf_size=16384
[2020/07/17 00:22:37] [debug] [router] default match rule stdin.0:stdout.0
[2020/07/17 00:22:37] [ info] [sp] stream processor started
[2020/07/17 00:22:37] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:22:37] [ warn] [input:stdin:stdin.0] end of file (stdin closed by remote end)
[2020/07/17 00:22:37] [ warn] [engine] service will stop in 5 seconds
[2020/07/17 00:22:42] [ info] [engine] service stopped
```

Custom buffer size:
```
[SERVICE]
    Log_Level       debug
    Daemon          Off
[INPUT]
    Name           stdin
    Buffer_size    1MB
[Output]
    Name           stdout
```
```
Fluent Bit v1.6.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/07/17 00:25:52] [ info] Configuration:
[2020/07/17 00:25:52] [ info]  flush time     | 5.000000 seconds
[2020/07/17 00:25:52] [ info]  grace          | 5 seconds
[2020/07/17 00:25:52] [ info]  daemon         | 0
[2020/07/17 00:25:52] [ info] ___________
[2020/07/17 00:25:52] [ info]  inputs:
[2020/07/17 00:25:52] [ info]      stdin
[2020/07/17 00:25:52] [ info] ___________
[2020/07/17 00:25:52] [ info]  filters:
[2020/07/17 00:25:52] [ info] ___________
[2020/07/17 00:25:52] [ info]  outputs:
[2020/07/17 00:25:52] [ info]      stdout.0
[2020/07/17 00:25:52] [ info] ___________
[2020/07/17 00:25:52] [ info]  collectors:
[2020/07/17 00:25:52] [ info] [engine] started (pid=2489192)
[2020/07/17 00:25:52] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/07/17 00:25:52] [debug] [storage] [cio stream] new stream registered: stdin.0
[2020/07/17 00:25:52] [ info] [storage] version=1.0.4, initializing...
[2020/07/17 00:25:52] [ info] [storage] in-memory
[2020/07/17 00:25:52] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/07/17 00:25:52] [debug] [input:stdin:stdin.0] buf_size=1000000
[2020/07/17 00:25:52] [debug] [router] default match rule stdin.0:stdout.0
[2020/07/17 00:25:52] [ info] [sp] stream processor started
[2020/07/17 00:25:52] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:25:52] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:25:52] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:25:52] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:25:52] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:25:52] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:25:52] [ warn] [input:stdin:stdin.0] end of file (stdin closed by remote end)
[2020/07/17 00:25:52] [debug] [task] created task=0x565193226110 id=0 OK
[2020/07/17 00:25:52] [ warn] [engine] service will stop in 5 seconds
[2020/07/17 00:25:52] [debug] [task] destroy task=0x565193226110 (task_id=0)
[2020/07/17 00:25:56] [ info] [engine] service stopped
```

**Comment by sxd** [comment]
Hi @dojci I've been taking a look into this issue for a few days now, and I have to say! thanks for the PR! please let me know when you think you're ready to take a deep look into!

For now I have a question, why you picked up 16Kb (the old value) and not increased the default to something like 1M ?
And another one, can you share the test case you're using to test?

Thanks for your work!

**Comment by dojci** [comment]
Valgrind output:
```
==2497479== Memcheck, a memory error detector
==2497479== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2497479== Using Valgrind-3.16.0.GIT and LibVEX; rerun with -h for copyright info
==2497479== Command: bin/fluent-bit -c /tmp/flb1.conf
==2497479== 
Fluent Bit v1.6.0
* Copyright (C) 2019-2020 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2020/07/17 00:27:27] [ info] Configuration:
[2020/07/17 00:27:27] [ info]  flush time     | 5.000000 seconds
[2020/07/17 00:27:27] [ info]  grace          | 5 seconds
[2020/07/17 00:27:27] [ info]  daemon         | 0
[2020/07/17 00:27:27] [ info] ___________
[2020/07/17 00:27:27] [ info]  inputs:
[2020/07/17 00:27:27] [ info]      stdin
[2020/07/17 00:27:27] [ info] ___________
[2020/07/17 00:27:27] [ info]  filters:
[2020/07/17 00:27:27] [ info] ___________
[2020/07/17 00:27:27] [ info]  outputs:
[2020/07/17 00:27:27] [ info]      stdout.0
[2020/07/17 00:27:27] [ info] ___________
[2020/07/17 00:27:27] [ info]  collectors:
[2020/07/17 00:27:27] [ info] [engine] started (pid=2497479)
[2020/07/17 00:27:27] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2020/07/17 00:27:27] [debug] [storage] [cio stream] new stream registered: stdin.0
[2020/07/17 00:27:27] [ info] [storage] version=1.0.4, initializing...
[2020/07/17 00:27:27] [ info] [storage] in-memory
[2020/07/17 00:27:27] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2020/07/17 00:27:27] [debug] [input:stdin:stdin.0] buf_size=1000000
[2020/07/17 00:27:27] [debug] [router] default match rule stdin.0:stdout.0
[2020/07/17 00:27:27] [ info] [sp] stream processor started
[2020/07/17 00:27:27] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:27:27] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:27:27] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:27:27] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:27:27] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:27:27] [debug] [input:stdin:stdin.0] data incomplete, waiting for more...
[2020/07/17 00:27:27] [ warn] [input:stdin:stdin.0] end of file (stdin closed by remote end)
[2020/07/17 00:27:27] [debug] [task] created task=0x51405f0 id=0 OK
[2020/07/17 00:27:27] [ warn] [engine] service will stop in 5 seconds
[2020/07/17 00:27:27] [debug] [task] destroy task=0x51405f0 (task_id=0)
[2020/07/17 00:27:32] [ info] [engine] service stopped
==2497479== 
==2497479== HEAP SUMMARY:
==2497479==     in use at exit: 0 bytes in 0 blocks
==2497479==   total heap usage: 253 allocs, 253 frees, 3,464,284 bytes allocated
==2497479== 
==2497479== All heap blocks were freed -- no leaks are possible
==2497479== 
==2497479== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by dojci** [comment]
> Hi @dojci I've been taking a look into this issue for a few days now, and I have to say! thanks for the PR! please let me know when you think you're ready to take a deep look into!
> 
> For now I have a question, why you picked up 16Kb (the old value) and not increased the default to something like 1M ?
> And another one, can you share the test case you're using to test?

Hi @sxd I usually try to keep the default values unchanged because various setups and components probably counts on that.
Test case is really simple:
```cat /tmp/test.json | bin/fluent-bit -c /tmp/flb.conf > /dev/null```

cat /tmp/flb.conf
```
[SERVICE]
    Log_Level       debug
    Daemon          Off
[INPUT]
    Name           stdin
    Buffer_size    1MB
[Output]
    Name           stdout
```
cat /tmp/test.json
```
{"message": "testtesttest...testtest"}
```
du -hs /tmp/test.json
```
436K   /tmp/test.json
```

**Comment by sxd** [comment]
@dojci Also, please don't forget to submit the proper PR in the documentation to add this new option

File: plugins/in_stdin/in_stdin.c [hunk 1]
```diff
@@ -210,45 +211,91 @@ static int in_stdin_collect(struct flb_input_instance *ins,
...
-    /* Allocate space for the configuration */
+    /* Allocate space for the configuration context */
     ctx = flb_malloc(sizeof(struct flb_in_stdin_config));
     if (!ctx) {
         return -1;
     }
+
+    ctx->buf = NULL;
     ctx->buf_len = 0;
     ctx->ins = in;
```
[the remark this answers is not in the document]
Comment (dojci) [reply plugins/in_stdin/in_stdin.c]: Fixed

File: plugins/in_stdin/in_stdin.c [hunk 2]
```diff
@@ -210,45 +211,91 @@ static int in_stdin_collect(struct flb_input_instance *ins,
...
+    }
+
+    /* buffer size setting */
+    pval = flb_input_get_property("buffer_size", in);
+    if (pval != NULL && flb_utils_size_to_bytes(pval) > 0) {
+        ctx->buf_size = flb_utils_size_to_bytes(pval);
+    }
+    else {
+        ctx->buf_size = DEFAULT_BUF_SIZE;
+    }
```
[the remark this answers is not in the document]
Comment (dojci) [reply plugins/in_stdin/in_stdin.c]: Yes it makes sense to have minimum value

File: plugins/in_stdin/in_stdin.c [hunk 3]
```diff
@@ -210,45 +211,91 @@ static int in_stdin_collect(struct flb_input_instance *ins,
...
+
+    ctx->buf = NULL;
     ctx->buf_len = 0;
     ctx->ins = in;
 
+    /* Initialize stdin config */
+    ret = in_stdin_config_init(ctx, in, config);
+    if (ret < 0) {
+        goto init_error;
+    }
```
[the remark this answers is not in the document]
Comment (dojci) [reply plugins/in_stdin/in_stdin.c:287]: I think you are right. Please see the latest changes.

File: plugins/in_stdin/in_stdin.c [hunk 4]
```diff
@@ -210,45 +211,96 @@ static int in_stdin_collect(struct flb_input_instance *ins,
...
+        ctx->buf_size = (size_t) flb_utils_size_to_bytes(pval);
+
+        if (ctx->buf_size == -1) {
+            flb_plg_error(ctx->ins, "buffer_size '%s' is invalid", pval);
+            return -1;
+        } else if (ctx->buf_size < DEFAULT_BUF_SIZE) {
+            flb_plg_error(ctx->ins, "buffer_size '%s' must be at least %i bytes",
+                          pval, DEFAULT_BUF_SIZE);
+            return -1;
+        }
```
Comment (dojci) [inline plugins/in_stdin/in_stdin.c:248]: any idea how to do this better ?

File: plugins/in_stdin/in_stdin.c [hunk 5]
```diff
@@ -210,45 +211,97 @@ static int in_stdin_collect(struct flb_input_instance *ins,
...
+
+        if (ctx->buf_size == -1) {
+            flb_plg_error(ctx->ins, "buffer_size '%s' is invalid", pval);
+            return -1;
+        }
+        else if (ctx->buf_size < DEFAULT_BUF_SIZE) {
+            flb_plg_error(ctx->ins, "buffer_size '%s' must be at least %i bytes",
+                          pval, DEFAULT_BUF_SIZE);
+            return -1;
+        }
```
Comment (dojci) [inline plugins/in_stdin/in_stdin.c:249]: @sxd any idea how to do this better ?

Comment (sxd) [reply plugins/in_stdin/in_stdin.c:249]: @dojci actually I think it's really easy to understand that piece of code, in my opinion it's fine and I'm ok with it

**Comment by sxd** [comment]
Looks pretty good to me, @edsiper can you take a look into this?
@dojci you still missing the documentation related to this option, please add it so people will know that this option exists.

**Comment by edsiper** [comment]
thanks!

note: likely the buffering option will be renamed before the release

**Pull request merged**